torch/quantization/fx/quantize.py

```diff
@@ -407,6 +414,687 @@ def load_arg(a):
+        assert is_activation_post_process_node(output_obs_node, modules)
+        parent_name, name = _parent_name(output_obs_node.target)
+        setattr(modules[parent_name], name, obs_mod_to_use)
+
+    # TODO(future PR): delete the orphaned observer modules
+
+def insert_observers_for_model_rewrite(
+        model: GraphModule,
+        modules: Dict[str, torch.nn.Module],
+        matches: Dict[str, MatchResult],
+        qconfig_map: Dict[str, QConfigAny],
+        observed_graph: Graph,
+        prepare_custom_config_dict: Dict[str, Any],
+        input_quantized_idxs: List[int],
+        output_quantized_idxs: List[int]) -> Optional[Node]:
+    """
+    Inserts observers, using the following high level algorithm:
+
+    For each node in the graph:
+      1. determine the target dtype of this node in the quantized graph, and save
+           it for future steps
+      2. determine the target dtype or all args and kwargs of this node
+      3. if any arg or kwarg's target dtype does not match the current node's
+           dtype, insert an observer
+      4. if the current node needs an output observer, insert it
+
+    For example:
+
+    - starting graph:
+        x0 -> linear -> x1
+
+    - observed graph after processing x0:
+        x0(fp32)
+
+    - observed graph after processing linear:
+        x0(fp32) -> x0_obs0(int8) -> linear(int8) -> linear_obs0(int8)
+
+    - observed graph after processing x1:
+        x0(fp32) -> x0_obs0(int8) -> linear(int8) -> linear_obs0(int8) -> x1
+
+    After a node is processed, the naive observer placement is guaranteed to be
+    complete for that node and all of its predecessors. There can be future
+    passes which optimize the graph by deduplicating observers, etc.
+    """
+
+    env: Dict[Any, Any] = {}
+    node_name_to_target_dtype: Dict[str, Any] = {}
+    cache_for_no_tensor_check: Dict[Node, bool] = dict()
+
+    def load_arg(a):
+        return map_arg(a, lambda node: env[node.name])
+
+    inputs_seen_counter = 0
+    outputs_seen_counter = 0
+    results_node = None
+
+    # first, populate the dtype map based only on qconfig and qhandler
+    # this assumes:
+    # graph inputs are fp32 by default, and int8 where overriden
+    # other nodes output dtype is specified by the qconfig
+    for node in model.graph.nodes:
+        root_node, matched_nodes, pattern, qhandler, qconfig = matches.get(
+            node.name, (None, None, None, None, None))
+        node_name_to_target_dtype[node.name] = get_target_dtype_for_node_rewrite(
+            node, qconfig, inputs_seen_counter, outputs_seen_counter,
+            input_quantized_idxs, output_quantized_idxs, qhandler)
+
+    # Second, for nodes with known input dtypes, propagate them throughout the
+    # graph. For example, if there is a call such as
+    #   x1 = x0.masked_fill(mask, 1)
+    # we propagate the type of mask to be torch.bool
+    propagate_dtypes_for_known_nodes_rewrite(
```

Comment:
should we incorporate all_node_args_have_no_tensors here as well? basically we are assuming all nodes are Tensors here, we can use None as a dtype for non-Tensors?